### PR TITLE
[OV]: load and convert llms in original precision

### PR DESCRIPTION
**Pull request by eaidova**
# What does this PR do?

allow loading bfloat16 and float16 models in original precision for conversion. It significantly reduces memory consumption and loading time during model conversion for large models

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/exporters/openvino/__main__.py [hunk 1]
```diff
@@ -84,6 +88,7 @@ def main_export(
     stateful: bool = True,
     convert_tokenizer: bool = False,
     library_name: Optional[str] = None,
+    model_loading_kwargs: Optional[Dict[str, Any]] = None,
```
[the remark this answers is not in the document]
Comment (eaidova) [reply optimum/exporters/openvino/__main__.py:91]: I thought about possibliity to force some other arguments in future e.g. attention_realization for switching between sdpa and eager manually, but it maybe too advanced, so for me having type argument is ok.

File: optimum/exporters/openvino/__main__.py [hunk 2]
```diff
@@ -361,6 +391,7 @@ class StoreAttr(object):
         preprocessors=preprocessors,
         device=device,
         trust_remote_code=trust_remote_code,
+        patch_16bit_model=patch_16bit,
```
Comment (echarlaix) [inline optimum/exporters/openvino/__main__.py:394]: could we instead update and use `ov_config.dtype` instead of adding `patch_16bit_model` ?

**Review by echarlaix — APPROVED** [review]
Looks great, thanks a lot @eaidova

File: optimum/exporters/openvino/__main__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/openvino/__main__.py:91]: thanks @eaidova !

File: optimum/exporters/openvino/__main__.py (hunk 2 above)
Comment (eaidova) [reply optimum/exporters/openvino/__main__.py:394]: layers patching is time consuming process we would like to avoid doing that for cases where it is not really neseccary. We need it only for pytorch model wegiths represented in fp16 and bf16 format for avoid issues tracing on cpu (torchscript does not support some ops running on cpu in this format). ov_config.dtype currently used for output model type and signal to compress weights to fp16 after conversion, mixing them in 1 parameter may lead to some issues:

* inability to convert fp32 pytorch models to fp16 openvino model (now it is possible)
* compressing model weights twice (having model in fp16 on torch side and then repeat compression for openvino)

**Comment by eaidova** [comment]
> Looks great, thanks a lot @eaidova

@echarlaix  thanks, we still investigating impact on models accuracy and quantization on our side. Could you please do not merge these changes, until we do not have whole picture?

File: optimum/exporters/openvino/__main__.py [hunk 3]
```diff
@@ -257,7 +266,28 @@ def main_export(
...
             trust_remote_code = False
+    dtype = loading_kwargs.get("torch_dtype")
+    if isinstance(dtype, str):
+        dtype = config.torch_dtype if dtype == "auto" else getattr(torch, dtype)
+
+    if (
+        dtype is None
+        and framework == "pt"
+        and not do_gptq_patching
+        and task.startswith("text-generation")
```
[the remark this answers is not in the document]
Comment (eaidova) [reply optimum/exporters/openvino/__main__.py:277]: in future we will propagate for other model types. text-generation models (especially LLMs) in most cases suffer for ineffective weights conversion when we load weights saved in float16/bfloat16 and implicit convert them to float32 when call from_pretrained method. The idea behind these changes is allowing load model as is for avoid weights conversion and also speedup some operations with weights like linear or matmul converting them directly to ov without passing tracing.

File: optimum/exporters/openvino/__main__.py [hunk 4]
```diff
@@ -213,6 +218,9 @@ def main_export(
         model_name_or_path, subfolder=subfolder, revision=revision, cache_dir=cache_dir, token=token
     )
 
+    if framework == "pt":
+        import torch
+
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/openvino/__main__.py:223]: I would prefer if imports are mostly at the top of the file, conditioned by the package's availability.

Comment (eaidova) [reply optimum/exporters/openvino/__main__.py:223]: I can not agree, it will lead to import some packages that can be not required during model export (e.g. import tf, if you export pytorch model) and leads to some failure that will block model export however these packages actually do not required to export model

File: optimum/exporters/openvino/__main__.py [hunk 5]
```diff
@@ -100,6 +104,7 @@ def main_export(
     stateful: bool = True,
     convert_tokenizer: bool = False,
     library_name: Optional[str] = None,
+    model_loading_kwargs: Optional[Dict[str, Any]] = None,
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/openvino/__main__.py:107]: why add a new argument instead of using `model_kwargs` which I believe is passed to the model's `from_pretrained`

File: optimum/exporters/openvino/__main__.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/openvino/__main__.py:223]: yes for tensorflow it's understandable, but torch is a required dependency in optimum either ways, and used at multiple stages of the export.

Comment (eaidova) [reply optimum/exporters/openvino/__main__.py]: ok, moved

**Review by IlyasMoutawwakil — APPROVED** [review]

**Comment by eaidova** [comment]
@IlyasMoutawwakil could you please merge?
